### PR TITLE
stats, executor: use a correct sampling to collect stats

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve?

Issue Number: close #xxx 

Problem Summary:

This pull request introduces [the weighted reservoir sampling](https://en.wikipedia.org/wiki/Reservoir_sampling#Algorithm_A-Res) with tests ensured.

And collect all statistics including the indexes' based on sampling.

It enables to maintain the multi-group stats which is not the prefix of some index to improve the estimation.

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

- Introduce the weighted reservoir sampling based on row(original one was based on column).
- Modifying unistore to support collect stats for sampling.
- Maintain the `AnalyzeExec` to support merge and collect the sampling from each region.
- Saving the sampling to storage.

### Related changes

TiKV PR link here
tipb PR link here

- PR to update `pingcap/docs`/`pingcap/docs-cn`:

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

- No release note

### CODE REVIEW

**Comment by winoros** [comment]
It will be split into smaller ones soon.

**Comment by time-and-fate** [comment]
I see the permission of some files changed (from 644 to 755) and there're some logs (seemingly) for debuging, they should be removed before merging.

**Comment by time-and-fate** [comment]
`TestVarsutil` needs to be modified to pass the CI.

File: planner/core/planbuilder.go [hunk 1]
```diff
@@ -1735,6 +1776,10 @@ func (b *PlanBuilder) buildAnalyzeIndex(as *ast.AnalyzeTableStmt, opts map[ast.A
 		}
 		b.ctx.GetSessionVars().StmtCtx.AppendWarning(errors.Errorf("The analyze version from the session is not compatible with the existing statistics of the table. Use the existing version instead"))
 	}
+	if version == statistics.Version3 {
+		b.ctx.GetSessionVars().StmtCtx.AppendWarning(errors.Errorf("The version 2 would collect all statistics not only the selected indexes"))
```
Comment (time-and-fate) [inline planner/core/planbuilder.go:1780]: version 2 => version 3

File: planner/core/planbuilder.go [hunk 2]
```diff
@@ -1796,6 +1841,10 @@ func (b *PlanBuilder) buildAnalyzeAllIndex(as *ast.AnalyzeTableStmt, opts map[as
...
```
Comment (time-and-fate) [inline planner/core/planbuilder.go:1845]: ditto.

**Comment by time-and-fate** [comment]
Do we still need to skip those tests since we've switched to version 3 now?

File: planner/core/planbuilder.go [hunk 3]
```diff
@@ -1638,6 +1638,43 @@ func GetPhysicalIDsAndPartitionNames(tblInfo *model.TableInfo, partitionNames []
...
+		info := analyzeInfo{
+			DBName:        tbl.Schema.O,
+			TableName:     tbl.Name.O,
+			PartitionName: names[i],
+			TableID:       AnalyzeTableID{TableID: tbl.TableInfo.ID, PartitionID: id},
+			Incremental:   as.Incremental,
+			StatsVersion:  version,
+		}
+		taskSlice = append(taskSlice, AnalyzeColumnsTask{
+			ColsInfo:    tbl.TableInfo.Columns,
```
[the remark this answers is not in the document]
Comment (winoros) [reply planner/core/planbuilder.go:1669]: It would be done in next pr.

**Comment by winoros** [comment]
> Do we still need to skip those tests since we've switched to version 3 now?

Removed

File: executor/builder.go [hunk 4]
```diff
@@ -2040,7 +2040,65 @@ func (b *executorBuilder) buildAnalyzeIndexIncremental(task plannercore.AnalyzeI
...
+		ColumnGroups: colGroups,
+	}
+	if task.TblInfo != nil {
+		e.analyzePB.ColReq.PrimaryColumnIds = tables.TryGetCommonPkColumnIds(task.TblInfo)
+		if task.TblInfo.IsCommonHandle {
+			e.analyzePB.ColReq.PrimaryPrefixColumnIds = tables.PrimaryPrefixColumnIDs(task.TblInfo)
+		}
+	}
+	b.err = plannercore.SetPBColumnsDefaultValue(b.ctx, e.analyzePB.ColReq.ColumnsInfo, task.ColsInfo)
+	job := &statistics.AnalyzeJob{DBName: task.DBName, TableName: task.TableName, PartitionName: task.PartitionName, JobInfo: autoAnalyze + "analyze columns"}
```
Comment (time-and-fate) [inline executor/builder.go:2094]: I think we should name the `JobInfo` "analyze table".

File: executor/analyze.go [hunk 5]
```diff
@@ -639,6 +672,129 @@ func (e *AnalyzeColumnsExec) buildResp(ranges []*ranger.Range) (distsql.SelectRe
...
+		}
+		colResp := &tipb.AnalyzeColumnsResp{}
+		err = colResp.Unmarshal(data)
+		if err != nil {
+			return 0, nil, nil, nil, err
+		}
+		subCollector := &statistics.RowSampleCollector{
+			MaxSampleSize: int(e.analyzePB.ColReq.SampleSize),
+		}
+		e.job.Update(subCollector.Count)
```
Comment (time-and-fate) [inline executor/analyze.go:718]: Seems we should use put his line below `subCollector.FromProto(colResp.RowCollector)`

**Review by time-and-fate — APPROVED** [review]
/lgtm

**Comment by time-and-fate** [comment]
I'll add correlation, extended stats calculation, and merge `BuildColumnHistAndTopN` and `BuildHistAndTopNOnRowSample` into one function in another PR.

File: executor/analyze.go [hunk 6]
```diff
@@ -517,6 +522,33 @@ func analyzeColumnsPushdown(colExec *AnalyzeColumnsExec) []analyzeResult {
...
+			TableID:  colExec.tableID,
+			Hist:     hists[cLen:],
+			TopNs:    topns[cLen:],
+			Fms:      fmSketches[cLen:],
+			job:      colExec.job,
+			StatsVer: colExec.analyzeVer,
+			Count:    count,
+			IsIndex:  1,
+		}
+		return []analyzeResult{colResult, idxResult}
```
Comment (eurekaka) [inline executor/analyze.go:550]: Return index stats in `analyzeColumnsPushdown` looks weird.

File: executor/builder.go [hunk 7]
```diff
@@ -2040,7 +2040,65 @@ func (b *executorBuilder) buildAnalyzeIndexIncremental(task plannercore.AnalyzeI
...
+		BucketSize:   int64(opts[ast.AnalyzeOptNumBuckets]),
+		SampleSize:   int64(opts[ast.AnalyzeOptNumSamples]),
+		SketchSize:   maxSketchSize,
+		ColumnsInfo:  util.ColumnsToProto(task.ColsInfo, task.TblInfo.PKIsHandle),
+		ColumnGroups: colGroups,
+	}
+	if task.TblInfo != nil {
+		e.analyzePB.ColReq.PrimaryColumnIds = tables.TryGetCommonPkColumnIds(task.TblInfo)
+		if task.TblInfo.IsCommonHandle {
+			e.analyzePB.ColReq.PrimaryPrefixColumnIds = tables.PrimaryPrefixColumnIDs(task.TblInfo)
```
Comment (eurekaka) [inline executor/builder.go:2090]: How is the `PrimaryPrefixColumnIds` used in coprocessor? why do we need to record it?

File: planner/core/planbuilder.go [hunk 8]
```diff
@@ -1730,6 +1771,10 @@ func (b *PlanBuilder) buildAnalyzeIndex(as *ast.AnalyzeTableStmt, opts map[ast.A
 		}
 		b.ctx.GetSessionVars().StmtCtx.AppendWarning(errors.Errorf("The analyze version from the session is not compatible with the existing statistics of the table. Use the existing version instead"))
 	}
+	if version == statistics.Version3 {
+		b.ctx.GetSessionVars().StmtCtx.AppendWarning(errors.Errorf("The version 3 would collect all statistics not only the selected indexes"))
+		return b.buildAnalyzeTable(as, opts, version)
+	}
```
Comment (eurekaka) [inline planner/core/planbuilder.go:1777]: Shall we mention that, `analyze index` is deprecated now?

File: statistics/row_sampler.go [hunk 9]
```diff
@@ -285,9 +289,161 @@ func RowSamplesToProto(samples WeightedRowSampleHeap) []*tipb.RowSample {
...
+			}
 			pbRow.Row = append(pbRow.Row, c.GetBytes())
 		}
 		rows = append(rows, pbRow)
 	}
 	return rows
 }
+
+// BuildHistAndTopNOnRowSample build a histogram and TopN for a column from samples.
+func BuildHistAndTopNOnRowSample(
```
Comment (eurekaka) [inline statistics/row_sampler.go:304]: Add a unit test for this function?

File: executor/analyze.go (hunk 6 above)
Comment (winoros) [reply executor/analyze.go:550]: Or we change the table to `analyzeTablePushDown`?

File: executor/builder.go (hunk 7 above)
Comment (winoros) [reply executor/builder.go:2090]: 😂 It's copied from original. Seems not used, let me check it again.

File: planner/core/planbuilder.go (hunk 8 above)
Comment (winoros) [reply planner/core/planbuilder.go:1777]: Let me add it.

File: statistics/row_sampler.go (hunk 9 above)
Comment (winoros) [reply statistics/row_sampler.go:304]: let me add it

File: executor/builder.go (hunk 7 above)
Comment (winoros) [reply executor/builder.go:2090]: It's used in unistore to decode row value.

File: executor/analyze.go (hunk 6 above)
Comment (winoros) [reply executor/analyze.go:550]: Changed the result to `colGroupResult`.

File: planner/core/planbuilder.go (hunk 8 above)
Comment (winoros) [reply planner/core/planbuilder.go:1777]: What warning message would be better?

**Review by eurekaka — COMMENTED** [review]
/lgtm

File: statistics/cmsketch.go [hunk 10]
```diff
@@ -487,6 +499,45 @@ type TopN struct {
...
+		if i+1 != len(c.TopN) {
+			fmt.Fprint(builder, ", ")
+		}
+	}
+	fmt.Fprint(builder, "]")
+	fmt.Fprint(builder, "}")
+	return builder.String()
+}
+
+func (c *TopN) DecodedString(ctx sessionctx.Context, colTypes []byte) (string, error) {
```
Comment (time-and-fate) [inline statistics/cmsketch.go:520]: This method needs comments to pass the linting.

File: statistics/cmsketch.go [hunk 11]
```diff
@@ -487,6 +499,45 @@ type TopN struct {
 	TopN []TopNMeta
 }
 
+func (c *TopN) String() string {
```
Comment (time-and-fate) [inline statistics/cmsketch.go:502]: Here also probably.

**Pull request merged**